From 9fb6dbe3c65fec7227a447647a34b3ec1445d498 Mon Sep 17 00:00:00 2001
From: Shubham Mittal <107728331+smittals2@users.noreply.github.com>
Date: Tue, 25 Feb 2025 11:40:27 -0800
Subject: [PATCH] Add SSL_CTX_use_cert_and_key   (#2163)

`CryptoAlg-2863`

Adds the function SSL_CTX_use_cert_and_key which sets the leaf, chain,
and private key all in one operation. This function is also needed by
librdkafka.

Regenerated error data

Point out areas that need special attention or support during the review
process. Discuss architecture or design changes.

Tests using this API to set the relevant fields and ensure TLS
connection succeeds.

By submitting this pull request, I confirm that my contribution is made
under the terms of the Apache 2.0 license and the ISC license.

---------

Co-authored-by: Justin W Smith <103147162+justsmth@users.noreply.github.com>
Co-authored-by: Justin Smith <justsmth@amazon.com>
---
 crypto/err/ssl.errordata |   1 +
 generated-src/err_data.c | 516 ++++++++++++++++++++-------------------
 include/openssl/ssl.h    |  19 ++
 ssl/internal.h           |   2 +
 ssl/ssl_cert.cc          | 131 ++++++++--
 ssl/ssl_session.cc       |   6 +-
 ssl/ssl_test.cc          |  42 ++++
 7 files changed, 436 insertions(+), 281 deletions(-)

diff --git a/crypto/err/ssl.errordata b/crypto/err/ssl.errordata
index 21734e357..35470a77d 100644
--- a/crypto/err/ssl.errordata
+++ b/crypto/err/ssl.errordata
@@ -112,6 +112,7 @@ SSL,308,NEGOTIATED_ALPS_WITHOUT_ALPN
 SSL,170,NEGOTIATED_BOTH_NPN_AND_ALPN
 SSL,285,NEGOTIATED_TB_WITHOUT_EMS_OR_RI
 SSL,171,NESTED_GROUP
+SSL,323,NOT_REPLACING_CERTIFICATE
 SSL,307,NO_APPLICATION_PROTOCOL
 SSL,172,NO_CERTIFICATES_RETURNED
 SSL,173,NO_CERTIFICATE_ASSIGNED
diff --git a/generated-src/err_data.c b/generated-src/err_data.c
index 96bca7dba..a01af0635 100644
--- a/generated-src/err_data.c
+++ b/generated-src/err_data.c
@@ -203,52 +203,52 @@ const uint32_t kOpenSSLReasonValues[] = {
     0x283500f7,
     0x28358d52,
     0x2836099a,
-    0x2c3237b6,
+    0x2c3237d0,
     0x2c32975c,
-    0x2c3337c4,
-    0x2c33b7d6,
-    0x2c3437ea,
-    0x2c34b7fc,
-    0x2c353817,
-    0x2c35b829,
-    0x2c363859,
+    0x2c3337de,
+    0x2c33b7f0,
+    0x2c343804,
+    0x2c34b816,
+    0x2c353831,
+    0x2c35b843,
+    0x2c363873,
     0x2c36833a,
-    0x2c373866,
-    0x2c37b892,
-    0x2c3838d0,
-    0x2c38b8e7,
-    0x2c393905,
-    0x2c39b915,
-    0x2c3a3927,
-    0x2c3ab93b,
-    0x2c3b394c,
-    0x2c3bb96b,
+    0x2c373880,
+    0x2c37b8ac,
+    0x2c3838ea,
+    0x2c38b901,
+    0x2c39391f,
+    0x2c39b92f,
+    0x2c3a3941,
+    0x2c3ab955,
+    0x2c3b3966,
+    0x2c3bb985,
     0x2c3c176e,
     0x2c3c9784,
-    0x2c3d39b0,
+    0x2c3d39ca,
     0x2c3d979d,
-    0x2c3e39da,
-    0x2c3eb9e8,
-    0x2c3f3a00,
-    0x2c3fba18,
-    0x2c403a42,
+    0x2c3e39f4,
+    0x2c3eba02,
+    0x2c3f3a1a,
+    0x2c3fba32,
+    0x2c403a5c,
     0x2c4093fd,
-    0x2c413a53,
-    0x2c41ba79,
+    0x2c413a6d,
+    0x2c41ba93,
     0x2c421382,
-    0x2c42ba8a,
+    0x2c42baa4,
     0x2c43076d,
-    0x2c43b95d,
-    0x2c4438a5,
-    0x2c44ba25,
-    0x2c45383c,
-    0x2c45b878,
-    0x2c4638f5,
-    0x2c46b97f,
-    0x2c473994,
-    0x2c47b9cd,
-    0x2c4838b7,
-    0x2c48ba66,
+    0x2c43b977,
+    0x2c4438bf,
+    0x2c44ba3f,
+    0x2c453856,
+    0x2c45b892,
+    0x2c46390f,
+    0x2c46b999,
+    0x2c4739ae,
+    0x2c47b9e7,
+    0x2c4838d1,
+    0x2c48ba80,
     0x30320000,
     0x30328015,
     0x3033001f,
@@ -459,146 +459,146 @@ const uint32_t kOpenSSLReasonValues[] = {
     0x4054a738,
     0x40552763,
     0x4055a7a0,
-    0x405627c5,
-    0x4056a7de,
-    0x405727f6,
-    0x4057a809,
-    0x4058281e,
-    0x4058a845,
-    0x40592874,
-    0x4059a8a1,
-    0x405aa8b5,
-    0x405b28cd,
-    0x405ba8de,
-    0x405c28f1,
-    0x405ca930,
-    0x405d293d,
-    0x405da962,
-    0x405e29a0,
+    0x405627df,
+    0x4056a7f8,
+    0x40572810,
+    0x4057a823,
+    0x40582838,
+    0x4058a85f,
+    0x4059288e,
+    0x4059a8bb,
+    0x405aa8cf,
+    0x405b28e7,
+    0x405ba8f8,
+    0x405c290b,
+    0x405ca94a,
+    0x405d2957,
+    0x405da97c,
+    0x405e29ba,
     0x405e8b2d,
-    0x405f29c1,
-    0x405fa9ce,
-    0x406029dc,
-    0x4060a9fe,
-    0x40612a5f,
-    0x4061aa97,
-    0x40622aae,
-    0x4062aabf,
-    0x40632b0c,
-    0x4063ab21,
-    0x40642b38,
-    0x4064ab64,
-    0x40652b7f,
-    0x4065ab96,
-    0x40662bae,
-    0x4066abd8,
-    0x40672c03,
-    0x4067ad06,
-    0x40682d4e,
-    0x4068ad6f,
-    0x40692da1,
-    0x4069adcf,
-    0x406a2df0,
-    0x406aae10,
-    0x406b2f98,
-    0x406bafbb,
-    0x406c2fd1,
-    0x406cb2db,
-    0x406d330a,
-    0x406db332,
-    0x406e3360,
-    0x406eb3ad,
-    0x406f3406,
-    0x406fb43e,
-    0x40703451,
-    0x4070b46e,
+    0x405f29db,
+    0x405fa9e8,
+    0x406029f6,
+    0x4060aa18,
+    0x40612a79,
+    0x4061aab1,
+    0x40622ac8,
+    0x4062aad9,
+    0x40632b26,
+    0x4063ab3b,
+    0x40642b52,
+    0x4064ab7e,
+    0x40652b99,
+    0x4065abb0,
+    0x40662bc8,
+    0x4066abf2,
+    0x40672c1d,
+    0x4067ad20,
+    0x40682d68,
+    0x4068ad89,
+    0x40692dbb,
+    0x4069ade9,
+    0x406a2e0a,
+    0x406aae2a,
+    0x406b2fb2,
+    0x406bafd5,
+    0x406c2feb,
+    0x406cb2f5,
+    0x406d3324,
+    0x406db34c,
+    0x406e337a,
+    0x406eb3c7,
+    0x406f3420,
+    0x406fb458,
+    0x4070346b,
+    0x4070b488,
     0x4071084d,
-    0x4071b480,
-    0x40723493,
-    0x4072b4c9,
-    0x407334e1,
+    0x4071b49a,
+    0x407234ad,
+    0x4072b4e3,
+    0x407334fb,
     0x40739986,
-    0x407434f5,
-    0x4074b50f,
-    0x40753520,
-    0x4075b534,
-    0x40763542,
+    0x4074350f,
+    0x4074b529,
+    0x4075353a,
+    0x4075b54e,
+    0x4076355c,
     0x40769732,
-    0x40773567,
-    0x4077b5a7,
-    0x407835c2,
-    0x4078b5fb,
-    0x40793612,
-    0x4079b628,
-    0x407a3654,
-    0x407ab667,
-    0x407b367c,
-    0x407bb68e,
-    0x407c36bf,
-    0x407cb6c8,
-    0x407d2d8a,
+    0x40773581,
+    0x4077b5c1,
+    0x407835dc,
+    0x4078b615,
+    0x4079362c,
+    0x4079b642,
+    0x407a366e,
+    0x407ab681,
+    0x407b3696,
+    0x407bb6a8,
+    0x407c36d9,
+    0x407cb6e2,
+    0x407d2da4,
     0x407da613,
-    0x407e35d7,
-    0x407ea855,
+    0x407e35f1,
+    0x407ea86f,
     0x407f2238,
     0x407fa40b,
     0x40802585,
     0x4080a260,
     0x408126c3,
     0x4081a512,
-    0x4082334b,
+    0x40823365,
     0x40829fb3,
-    0x40832830,
-    0x4083ab49,
+    0x4083284a,
+    0x4083ab63,
     0x40842274,
-    0x4084a88d,
-    0x40852902,
-    0x4085aa26,
-    0x40862982,
+    0x4084a8a7,
+    0x4085291c,
+    0x4085aa40,
+    0x4086299c,
     0x4086a62d,
-    0x40873391,
-    0x4087aa74,
+    0x408733ab,
+    0x4087aa8e,
     0x40881ff1,
-    0x4088ad19,
+    0x4088ad33,
     0x40892040,
     0x40899fcd,
-    0x408a3009,
+    0x408a3023,
     0x408a9d9e,
-    0x408b36a3,
-    0x408bb41b,
-    0x408c2912,
+    0x408b36bd,
+    0x408bb435,
+    0x408c292c,
     0x408c9dd6,
     0x408d235c,
     0x408da2a6,
     0x408e248c,
     0x408ea780,
-    0x408f2d2d,
-    0x408faa42,
-    0x40902c24,
-    0x4090a954,
-    0x40912ff1,
+    0x408f2d47,
+    0x408faa5c,
+    0x40902c3e,
+    0x4090a96e,
+    0x4091300b,
     0x40919dfc,
     0x4092208d,
-    0x4092b3cc,
-    0x409334ac,
+    0x4092b3e6,
+    0x409334c6,
     0x4093a63e,
     0x40942288,
-    0x4094b022,
-    0x40952ad0,
-    0x4095b634,
-    0x40963378,
+    0x4094b03c,
+    0x40952aea,
+    0x4095b64e,
+    0x40963392,
     0x4096a59e,
     0x40972689,
     0x4097a4db,
     0x409820ed,
-    0x4098aae4,
-    0x409933e8,
-    0x4099a7ad,
+    0x4098aafe,
+    0x40993402,
+    0x4099a7c7,
     0x409a2746,
     0x409a9dba,
     0x409b22e2,
     0x409ba30d,
-    0x409c3589,
+    0x409c35a3,
     0x409ca335,
     0x409d255a,
     0x409da528,
@@ -609,48 +609,49 @@ const uint32_t kOpenSSLReasonValues[] = {
     0x40a025fb,
     0x40a0a4f5,
     0x40a12543,
-    0x40fa2cec,
-    0x40faac48,
-    0x40fb2ccb,
-    0x40fbac62,
-    0x40fcacaa,
-    0x40fd2c83,
+    0x40a1a7ad,
+    0x40fa2d06,
+    0x40faac62,
+    0x40fb2ce5,
+    0x40fbac7c,
+    0x40fcacc4,
+    0x40fd2c9d,
     0x40fd9efe,
     0x40fe1f12,
-    0x41f42ec3,
-    0x41f92f55,
-    0x41fe2e48,
-    0x41feb0fe,
-    0x41ff322c,
-    0x42032edc,
-    0x42082efe,
-    0x4208af3a,
-    0x42092e2c,
-    0x4209af74,
-    0x420a2e83,
-    0x420aae63,
-    0x420b2ea3,
-    0x420baf1c,
-    0x420c3248,
-    0x420cb032,
-    0x420d30e5,
-    0x420db11c,
-    0x4212314f,
-    0x4217320f,
-    0x4217b191,
-    0x421c31b3,
-    0x421f316e,
-    0x422132c0,
-    0x422631f2,
-    0x422b329e,
-    0x422bb0c0,
-    0x422c3280,
-    0x422cb073,
-    0x422d304c,
-    0x422db25f,
-    0x422e309f,
-    0x423031ce,
-    0x4230b136,
+    0x41f42edd,
+    0x41f92f6f,
+    0x41fe2e62,
+    0x41feb118,
+    0x41ff3246,
+    0x42032ef6,
+    0x42082f18,
+    0x4208af54,
+    0x42092e46,
+    0x4209af8e,
+    0x420a2e9d,
+    0x420aae7d,
+    0x420b2ebd,
+    0x420baf36,
+    0x420c3262,
+    0x420cb04c,
+    0x420d30ff,
+    0x420db136,
+    0x42123169,
+    0x42173229,
+    0x4217b1ab,
+    0x421c31cd,
+    0x421f3188,
+    0x422132da,
+    0x4226320c,
+    0x422b32b8,
+    0x422bb0da,
+    0x422c329a,
+    0x422cb08d,
+    0x422d3066,
+    0x422db279,
+    0x422e30b9,
+    0x423031e8,
+    0x4230b150,
     0x42310b85,
     0x44320778,
     0x44328787,
@@ -707,71 +708,71 @@ const uint32_t kOpenSSLReasonValues[] = {
     0x4c419866,
     0x4c4219cf,
     0x4c4297ae,
-    0x50323a9c,
-    0x5032baab,
-    0x50333ab6,
-    0x5033bac6,
-    0x50343adf,
-    0x5034baf9,
-    0x50353b07,
-    0x5035bb1d,
-    0x50363b2f,
-    0x5036bb45,
-    0x50373b5e,
-    0x5037bb71,
-    0x50383b89,
-    0x5038bb9a,
-    0x50393baf,
-    0x5039bbc3,
-    0x503a3be3,
-    0x503abbf9,
-    0x503b3c11,
-    0x503bbc23,
-    0x503c3c3f,
-    0x503cbc56,
-    0x503d3c6f,
-    0x503dbc85,
-    0x503e3c92,
-    0x503ebca8,
-    0x503f3cba,
+    0x50323ab6,
+    0x5032bac5,
+    0x50333ad0,
+    0x5033bae0,
+    0x50343af9,
+    0x5034bb13,
+    0x50353b21,
+    0x5035bb37,
+    0x50363b49,
+    0x5036bb5f,
+    0x50373b78,
+    0x5037bb8b,
+    0x50383ba3,
+    0x5038bbb4,
+    0x50393bc9,
+    0x5039bbdd,
+    0x503a3bfd,
+    0x503abc13,
+    0x503b3c2b,
+    0x503bbc3d,
+    0x503c3c59,
+    0x503cbc70,
+    0x503d3c89,
+    0x503dbc9f,
+    0x503e3cac,
+    0x503ebcc2,
+    0x503f3cd4,
     0x503f83b3,
-    0x50403ccd,
-    0x5040bcdd,
-    0x50413cf7,
-    0x5041bd06,
-    0x50423d20,
-    0x5042bd3d,
-    0x50433d4d,
-    0x5043bd5d,
-    0x50443d7a,
+    0x50403ce7,
+    0x5040bcf7,
+    0x50413d11,
+    0x5041bd20,
+    0x50423d3a,
+    0x5042bd57,
+    0x50433d67,
+    0x5043bd77,
+    0x50443d94,
     0x50448469,
-    0x50453d8e,
-    0x5045bdac,
-    0x50463dbf,
-    0x5046bdd5,
-    0x50473de7,
-    0x5047bdfc,
-    0x50483e22,
-    0x5048be30,
-    0x50493e43,
-    0x5049be58,
-    0x504a3e6e,
-    0x504abe7e,
-    0x504b3e9e,
-    0x504bbeb1,
-    0x504c3ed4,
-    0x504cbf02,
-    0x504d3f2f,
-    0x504dbf4c,
-    0x504e3f67,
-    0x504ebf83,
-    0x504f3f95,
-    0x504fbfac,
-    0x50503fbb,
+    0x50453da8,
+    0x5045bdc6,
+    0x50463dd9,
+    0x5046bdef,
+    0x50473e01,
+    0x5047be16,
+    0x50483e3c,
+    0x5048be4a,
+    0x50493e5d,
+    0x5049be72,
+    0x504a3e88,
+    0x504abe98,
+    0x504b3eb8,
+    0x504bbecb,
+    0x504c3eee,
+    0x504cbf1c,
+    0x504d3f49,
+    0x504dbf66,
+    0x504e3f81,
+    0x504ebf9d,
+    0x504f3faf,
+    0x504fbfc6,
+    0x50503fd5,
     0x50508729,
-    0x50513fce,
-    0x5051bd6c,
-    0x50523f14,
+    0x50513fe8,
+    0x5051bd86,
+    0x50523f2e,
     0x583210b7,
     0x5c329409,
     0x5c331422,
@@ -855,19 +856,19 @@ const uint32_t kOpenSSLReasonValues[] = {
     0x7c321398,
     0x80321879,
     0x80328090,
-    0x80333785,
+    0x8033379f,
     0x803380b9,
-    0x80343794,
-    0x8034b6fc,
-    0x8035371a,
-    0x8035b7a8,
-    0x8036375c,
-    0x8036b70b,
-    0x8037374e,
-    0x8037b6e9,
-    0x8038376f,
-    0x8038b72b,
-    0x80393740,
+    0x803437ae,
+    0x8034b716,
+    0x80353734,
+    0x8035b7c2,
+    0x80363776,
+    0x8036b725,
+    0x80373768,
+    0x8037b703,
+    0x80383789,
+    0x8038b745,
+    0x8039375a,
 };
 
 const size_t kOpenSSLReasonValuesLen = sizeof(kOpenSSLReasonValues) / sizeof(kOpenSSLReasonValues[0]);
@@ -1359,6 +1360,7 @@ const char kOpenSSLReasonStringData[] =
     "NEGOTIATED_BOTH_NPN_AND_ALPN\0"
     "NEGOTIATED_TB_WITHOUT_EMS_OR_RI\0"
     "NESTED_GROUP\0"
+    "NOT_REPLACING_CERTIFICATE\0"
     "NO_APPLICATION_PROTOCOL\0"
     "NO_CERTIFICATES_RETURNED\0"
     "NO_CERTIFICATE_ASSIGNED\0"
diff --git a/include/openssl/ssl.h b/include/openssl/ssl.h
index daf02bf5e..dfbea1ee7 100644
--- a/include/openssl/ssl.h
+++ b/include/openssl/ssl.h
@@ -1257,6 +1257,24 @@ OPENSSL_EXPORT int SSL_set_signing_algorithm_prefs(SSL *ssl,
                                                    const uint16_t *prefs,
                                                    size_t num_prefs);
 
+// SSL_CTX_use_cert_and_key sets |x509|, |privatekey|, and |chain| on |ctx|.
+// The |privatekey| argument must be the private key of the certificate |x509|.
+// If the override argument is 0, then |x509|, |privatekey|, and |chain| are
+// set only if all were not previously set. If override is non-0, then the
+// certificate, private key and chain certs are always set. |privatekey| and
+// |x509| are not copied or duplicated, their reference counts are incremented.
+// In OpenSSL, a shallow copy of |chain| is stored with a reference count
+// increment for all |X509| objects in the chain. In AWS-LC,
+// we represent X509 chains as a CRYPTO_BUFFER stack. Therefore, we create a
+// an internal copy and leave the |chain| parameter untouched. This means,
+// changes to |chain| after this function is called will not update in |ctx|.
+// This is different from OpenSSL which stores a reference to the X509
+// certificates in the |chain| object.
+//
+// Returns one on success and zero on error.
+OPENSSL_EXPORT int SSL_CTX_use_cert_and_key(SSL_CTX *ctx, X509 *x509,
+                                            EVP_PKEY *privatekey,
+                                            STACK_OF(X509) *chain, int override);
 
 // Certificate and private key convenience functions.
 
@@ -6460,6 +6478,7 @@ BSSL_NAMESPACE_END
 #define SSL_R_INVALID_OUTER_EXTENSION 320
 #define SSL_R_INCONSISTENT_ECH_NEGOTIATION 321
 #define SSL_R_INVALID_ALPS_CODEPOINT 322
+#define SSL_R_NOT_REPLACING_CERTIFICATE 323
 #define SSL_R_SERIALIZATION_UNSUPPORTED 500
 #define SSL_R_SERIALIZATION_INVALID_SSL 501
 #define SSL_R_SERIALIZATION_INVALID_SSL_CONFIG 502
diff --git a/ssl/internal.h b/ssl/internal.h
index 38617640c..b81f73394 100644
--- a/ssl/internal.h
+++ b/ssl/internal.h
@@ -3350,6 +3350,8 @@ bool ssl_compare_public_and_private_key(const EVP_PKEY *pubkey,
                                         const EVP_PKEY *privkey);
 bool ssl_cert_check_private_key(const CERT *cert, const EVP_PKEY *privkey);
 
+CRYPTO_BUFFER *buffer_up_ref(const CRYPTO_BUFFER *buffer);
+
 // ssl_cert_check_cert_private_keys_usage returns true if |cert_private_keys|
 // in |cert| has a valid index and a sufficient amount of slots.
 bool ssl_cert_check_cert_private_keys_usage(const CERT *cert);
diff --git a/ssl/ssl_cert.cc b/ssl/ssl_cert.cc
index 6a4b63c54..b6f04a07e 100644
--- a/ssl/ssl_cert.cc
+++ b/ssl/ssl_cert.cc
@@ -145,7 +145,7 @@ CERT::CERT(const SSL_X509_METHOD *x509_method_arg)
 
 CERT::~CERT() { x509_method->cert_free(this); }
 
-static CRYPTO_BUFFER *buffer_up_ref(const CRYPTO_BUFFER *buffer) {
+CRYPTO_BUFFER *buffer_up_ref(const CRYPTO_BUFFER *buffer) {
   CRYPTO_BUFFER_up_ref(const_cast<CRYPTO_BUFFER *>(buffer));
   return const_cast<CRYPTO_BUFFER *>(buffer);
 }
@@ -274,7 +274,7 @@ static enum leaf_cert_and_privkey_result_t check_leaf_cert_and_privkey(
 }
 
 static int cert_set_chain_and_key(
-    CERT *cert, CRYPTO_BUFFER *const *certs, size_t num_certs,
+    CERT *cert, UniquePtr<STACK_OF(CRYPTO_BUFFER)> &certs, size_t num_certs,
     EVP_PKEY *privkey, const SSL_PRIVATE_KEY_METHOD *privkey_method) {
   if (num_certs == 0 || (privkey == NULL && privkey_method == NULL)) {
     OPENSSL_PUT_ERROR(SSL, ERR_R_PASSED_NULL_PARAMETER);
@@ -286,7 +286,7 @@ static int cert_set_chain_and_key(
     return 0;
   }
 
-  switch (check_leaf_cert_and_privkey(certs[0], privkey)) {
+  switch (check_leaf_cert_and_privkey(sk_CRYPTO_BUFFER_value(certs.get(), 0), privkey)) {
     case leaf_cert_and_privkey_error:
       return 0;
     case leaf_cert_and_privkey_mismatch:
@@ -296,27 +296,25 @@ static int cert_set_chain_and_key(
       break;
   }
 
-  UniquePtr<STACK_OF(CRYPTO_BUFFER)> certs_sk(sk_CRYPTO_BUFFER_new_null());
-  if (!certs_sk) {
-    return 0;
-  }
-
-  for (size_t i = 0; i < num_certs; i++) {
-    if (!PushToStack(certs_sk.get(), UpRef(certs[i]))) {
-      return 0;
-    }
-  }
-
   if (!ssl_cert_check_cert_private_keys_usage(cert)) {
     return 0;
   }
 
-  // Update certificate slot index once all checks have passed.
   // Certificate slot validity already checked in |check_leaf_cert_and_privkey|.
   int idx = ssl_get_certificate_slot_index(privkey);
-  cert->cert_private_keys[idx].privatekey = UpRef(privkey);
+  assert(idx >= 0);
+  CERT_PKEY *cert_pkey = &cert->cert_private_keys[idx];
+
+  // Update certificate slot index once all checks have passed.
+  if (cert_pkey->privatekey) {
+    cert_pkey->privatekey.reset();
+  }
+  cert_pkey->privatekey = UpRef(privkey);
   cert->key_method = privkey_method;
-  cert->cert_private_keys[idx].chain = std::move(certs_sk);
+  if (cert_pkey->chain) {
+    cert_pkey->chain.reset();
+  }
+  cert_pkey->chain = std::move(certs);
   cert->cert_private_key_idx = idx;
   return 1;
 }
@@ -944,23 +942,118 @@ BSSL_NAMESPACE_END
 
 using namespace bssl;
 
+static int cert_array_to_stack(CRYPTO_BUFFER *const *from,
+                              UniquePtr<STACK_OF(CRYPTO_BUFFER)> *to,
+                              size_t num_certs) {
+  for (size_t i = 0; i < num_certs; i++) {
+    if (!PushToStack(to->get(), UpRef(from[i]))) {
+      return 0;
+    }
+  }
+  return 1;
+}
+
 int SSL_set_chain_and_key(SSL *ssl, CRYPTO_BUFFER *const *certs,
                           size_t num_certs, EVP_PKEY *privkey,
                           const SSL_PRIVATE_KEY_METHOD *privkey_method) {
   if (!ssl->config) {
     return 0;
   }
-  return cert_set_chain_and_key(ssl->config->cert.get(), certs, num_certs,
+  UniquePtr<STACK_OF(CRYPTO_BUFFER)> crypto_certs(sk_CRYPTO_BUFFER_new_null());
+  if (!cert_array_to_stack(certs, &crypto_certs, num_certs)) {
+    return 0;
+  }
+  return cert_set_chain_and_key(ssl->config->cert.get(), crypto_certs, num_certs,
                                 privkey, privkey_method);
 }
 
 int SSL_CTX_set_chain_and_key(SSL_CTX *ctx, CRYPTO_BUFFER *const *certs,
                               size_t num_certs, EVP_PKEY *privkey,
                               const SSL_PRIVATE_KEY_METHOD *privkey_method) {
-  return cert_set_chain_and_key(ctx->cert.get(), certs, num_certs, privkey,
+  UniquePtr<STACK_OF(CRYPTO_BUFFER)> crypto_certs(sk_CRYPTO_BUFFER_new_null());
+  if (!cert_array_to_stack(certs, &crypto_certs, num_certs)) {
+    return 0;
+  }
+  return cert_set_chain_and_key(ctx->cert.get(), crypto_certs, num_certs, privkey,
                                 privkey_method);
 }
 
+int SSL_CTX_use_cert_and_key(SSL_CTX *ctx, X509 *x509, EVP_PKEY *privatekey,
+                             STACK_OF(X509) *chain, int override) {
+
+  if (privatekey == nullptr || x509 == nullptr) {
+    OPENSSL_PUT_ERROR(SSL, ERR_R_PASSED_NULL_PARAMETER);
+    return 0;
+  }
+
+  // Check override value
+  CERT *cert = ctx->cert.get();
+  int idx = ssl_get_certificate_slot_index(privatekey);
+  if (idx < 0) {
+    OPENSSL_PUT_ERROR(SSL, SSL_R_UNKNOWN_CERTIFICATE_TYPE);
+    return 0;
+  }
+  CERT_PKEY *cert_pkey = &cert->cert_private_keys[idx];
+  if (!override && (cert_pkey->privatekey != NULL ||
+      cert_pkey->x509_leaf != NULL ||
+      cert_pkey->chain != NULL)) {
+    OPENSSL_PUT_ERROR(SSL, SSL_R_NOT_REPLACING_CERTIFICATE);
+    return 0;
+  }
+
+  UniquePtr<STACK_OF(CRYPTO_BUFFER)> leaf_and_chain(sk_CRYPTO_BUFFER_new_null());
+
+  // Convert leaf certificate |x509| to CRYPTO_BUFFER
+  uint8_t *buf = nullptr;
+  int cert_len = i2d_X509(x509, &buf);
+  if (cert_len <= 0) {
+    OPENSSL_free(buf);
+    return 0;
+  }
+
+  // Add leaf cert first to the chain
+  UniquePtr<CRYPTO_BUFFER> leaf_buf(CRYPTO_BUFFER_new(buf, cert_len, nullptr));
+  OPENSSL_free(buf);
+  if (!leaf_buf ||
+      !PushToStack(leaf_and_chain.get(), std::move(leaf_buf))) {
+    return 0;
+  }
+
+  // Convert other chain certificates to CRYPTO_BUFFER objects
+  if (chain != nullptr) {
+    for (size_t i = 0; i < sk_X509_num(chain); i++) {
+      buf = nullptr;
+      cert_len = i2d_X509(sk_X509_value(chain, i), &buf);
+      if (cert_len <= 0) {
+        OPENSSL_free(buf);
+        return 0;
+      }
+
+      UniquePtr<CRYPTO_BUFFER> chain_buf(CRYPTO_BUFFER_new(buf, cert_len, nullptr));
+      OPENSSL_free(buf);
+      if (!chain_buf ||
+          !PushToStack(leaf_and_chain.get(), std::move(chain_buf))) {
+        return 0;
+      }
+    }
+  }
+
+  // Call SSL_CTX_set_chain_and_key to set the chain and key
+  if (!cert_set_chain_and_key(cert, leaf_and_chain,
+                              sk_CRYPTO_BUFFER_num(leaf_and_chain.get()),
+                              privatekey, nullptr)) {
+    return 0;
+  }
+
+  // Update the leaf certificate
+  if (cert_pkey->x509_leaf) {
+    X509_free(cert_pkey->x509_leaf);
+  }
+  X509_up_ref(x509);
+  cert_pkey->x509_leaf = x509;
+  return 1;
+}
+
 void SSL_certs_clear(SSL *ssl) {
   if (!ssl->config) {
     return;
diff --git a/ssl/ssl_session.cc b/ssl/ssl_session.cc
index c9116d01b..0cff79a69 100644
--- a/ssl/ssl_session.cc
+++ b/ssl/ssl_session.cc
@@ -214,12 +214,8 @@ UniquePtr<SSL_SESSION> SSL_SESSION_dup(SSL_SESSION *session, int dup_flags) {
     }
   }
   if (session->certs != nullptr) {
-    auto buf_up_ref = [](const CRYPTO_BUFFER *buf) {
-      CRYPTO_BUFFER_up_ref(const_cast<CRYPTO_BUFFER *>(buf));
-      return const_cast<CRYPTO_BUFFER*>(buf);
-    };
     new_session->certs.reset(sk_CRYPTO_BUFFER_deep_copy(
-        session->certs.get(), buf_up_ref, CRYPTO_BUFFER_free));
+        session->certs.get(), buffer_up_ref, CRYPTO_BUFFER_free));
     if (new_session->certs == nullptr) {
       return nullptr;
     }
diff --git a/ssl/ssl_test.cc b/ssl/ssl_test.cc
index 2c4b82483..50b931313 100644
--- a/ssl/ssl_test.cc
+++ b/ssl/ssl_test.cc
@@ -6216,6 +6216,12 @@ TEST(SSLTest, SetChainAndKeyMismatch) {
   ASSERT_FALSE(SSL_CTX_set_chain_and_key(ctx.get(), &chain[0], chain.size(),
                                          key.get(), nullptr));
   ERR_clear_error();
+
+  // Ensure |SSL_CTX_use_cert_and_key| also fails
+  bssl::UniquePtr<X509> x509_leaf = X509FromBuffer(GetChainTestCertificateBuffer());
+  ASSERT_FALSE(SSL_CTX_use_cert_and_key(ctx.get(), x509_leaf.get(),
+                                        key.get(), NULL, 1));
+  ERR_clear_error();
 }
 
 TEST(SSLTest, SetChainAndKey) {
@@ -6254,6 +6260,42 @@ TEST(SSLTest, SetChainAndKey) {
                                      server_ctx.get()));
 }
 
+TEST(SSLTest, SetLeafChainAndKey) {
+  bssl::UniquePtr<SSL_CTX> client_ctx(SSL_CTX_new(TLS_method()));
+  ASSERT_TRUE(client_ctx);
+  bssl::UniquePtr<SSL_CTX> server_ctx(SSL_CTX_new(TLS_method()));
+  ASSERT_TRUE(server_ctx);
+
+  ASSERT_EQ(nullptr, SSL_CTX_get0_chain(server_ctx.get()));
+
+  bssl::UniquePtr<EVP_PKEY> key = GetChainTestKey();
+  ASSERT_TRUE(key);
+  bssl::UniquePtr<X509> leaf = GetChainTestCertificate();
+  ASSERT_TRUE(leaf);
+  bssl::UniquePtr<X509> intermediate = GetChainTestIntermediate();
+  bssl::UniquePtr<STACK_OF(X509)> chain(sk_X509_new_null());
+  ASSERT_TRUE(chain);
+  ASSERT_TRUE(PushToStack(chain.get(), std::move(intermediate)));
+
+  ASSERT_TRUE(SSL_CTX_use_cert_and_key(server_ctx.get(), leaf.get(),
+                                        key.get(), chain.get(), 1));
+
+  SSL_CTX_set_custom_verify(
+      client_ctx.get(), SSL_VERIFY_PEER,
+      [](SSL *ssl, uint8_t *out_alert) -> ssl_verify_result_t {
+        return ssl_verify_ok;
+      });
+
+  bssl::UniquePtr<SSL> client, server;
+  ASSERT_TRUE(ConnectClientAndServer(&client, &server, client_ctx.get(),
+                                     server_ctx.get()));
+
+  // Try setting on previously populated fields without an override
+  ASSERT_FALSE(SSL_CTX_use_cert_and_key(server_ctx.get(), leaf.get(),
+                                        key.get(), chain.get(), 0));
+  ERR_clear_error();
+}
+
 TEST(SSLTest, BuffersFailWithoutCustomVerify) {
   bssl::UniquePtr<SSL_CTX> client_ctx(SSL_CTX_new(TLS_with_buffers_method()));
   ASSERT_TRUE(client_ctx);
-- 
2.47.0

